### PR TITLE
Added support for nullable values (fixes #152)

### LINKED ISSUE #152: Nullable string serialized as double-quoted empty string
### Expected Behavior
As mentionned in the docs (https://github.com/fastify/fast-json-stringify#nullable), properties marked as `nullable: true` should be serialized as `null`.

### Current Behavior
Currently, a null string is serialized as a double-quoted empty string (```""```)

### Steps to Reproduce
```js
const fastJson = require('fast-json-stringify')

let stringify = fastJson({
  type: 'object',
  properties: {
    nullableString: { type: 'string', nullable: true },
  }
})

let data = {
  nullableString: null
}

console.log(stringify(data))

// => {"nullableString":""}
```

### Workaround
Remove `nullable: true`  and mark property type as `['string', 'null']`

### Context (Environment)
- node v11.5.0
- fast-json-stringify v1.11.3

### PR DESCRIPTION
**Pull request by gendronb**

### CODE REVIEW

**Comment by mcollina** [comment]
Good work!

**Comment by gendronb** [comment]
> Can you check if this impacts our benchmarks anyhow?

@mcollina I ran the benchmarks, here are my numbers (not sure how to interpret these) : 

```
$ node bench.js
FJS creation x 4,092 ops/sec ±6.19% (83 runs sampled)
JSON.stringify array x 2,065 ops/sec ±3.90% (86 runs sampled)
fast-json-stringify array x 4,701 ops/sec ±2.37% (88 runs sampled)
fast-json-stringify-uglified array x 4,584 ops/sec ±3.29% (86 runs sampled)
JSON.stringify long string x 9,540 ops/sec ±2.09% (88 runs sampled)
fast-json-stringify long string x 7,328 ops/sec ±10.83% (70 runs sampled)
fast-json-stringify-uglified long string x 6,406 ops/sec ±17.36% (61 runs sampled)
JSON.stringify short string x 3,484,012 ops/sec ±5.76% (82 runs sampled)
fast-json-stringify short string x 17,517,700 ops/sec ±3.36% (84 runs sampled)
fast-json-stringify-uglified short string x 17,451,596 ops/sec ±3.00% (82 runs sampled)
JSON.stringify obj x 901,404 ops/sec ±21.27% (69 runs sampled)
fast-json-stringify obj x 3,065,142 ops/sec ±20.89% (71 runs sampled)
fast-json-stringify-uglified obj x 4,199,860 ops/sec ±7.61% (83 runs sampled)
Done in 73.04s.
```

**Comment by mcollina** [comment]
You should also run them on master

**Comment by gendronb** [comment]
> You should also run them on master

```
Bruno@bruno-laptop MINGW64 /c/Data/dev/02-forks/fast-json-stringify (master)
$ yarn benchmark
yarn run v1.12.3
$ node bench.js
FJS creation x 4,092 ops/sec ±6.19% (83 runs sampled)
JSON.stringify array x 2,065 ops/sec ±3.90% (86 runs sampled)
fast-json-stringify array x 4,701 ops/sec ±2.37% (88 runs sampled)
fast-json-stringify-uglified array x 4,584 ops/sec ±3.29% (86 runs sampled)
JSON.stringify long string x 9,540 ops/sec ±2.09% (88 runs sampled)
fast-json-stringify long string x 7,328 ops/sec ±10.83% (70 runs sampled)
fast-json-stringify-uglified long string x 6,406 ops/sec ±17.36% (61 runs sampled)
JSON.stringify short string x 3,484,012 ops/sec ±5.76% (82 runs sampled)
fast-json-stringify short string x 17,517,700 ops/sec ±3.36% (84 runs sampled)
fast-json-stringify-uglified short string x 17,451,596 ops/sec ±3.00% (82 runs sampled)
JSON.stringify obj x 901,404 ops/sec ±21.27% (69 runs sampled)
fast-json-stringify obj x 3,065,142 ops/sec ±20.89% (71 runs sampled)
fast-json-stringify-uglified obj x 4,199,860 ops/sec ±7.61% (83 runs sampled)
Done in 73.04s.
```

**Comment by mcollina** [comment]
It seems that with this fix we are losing 30% of throughput when stringifying objects. I think this is due to the added functions. Can you implement this feature without adding a function for every property?

**Comment by gendronb** [comment]
@mcollina yeah, that's what I was afraid of...

Ok, let me have another look at this, I will probably add a check before adding each function, but this will be at the expense of speed when building the schema...

**Comment by mcollina** [comment]
That's ok. The priority is to get speedy at runtime.

Note that https://github.com/fastify/fast-json-stringify/pull/153/files#diff-168726dbe96b3ce427e7fedce31bb0bcR77 can be checked at compile time, and then you can avoid adding the function.

**Comment by gendronb** [comment]
That's the way i'm heading at the moment, conditionnally adding functions at compile time...

Stay tuned!

**Comment by gendronb** [comment]
And removing all check for nullable inside functions...

**Comment by mcollina** [comment]
How does this affect benchmarks?

**Comment by gendronb** [comment]
@mcollina please find the results below... 

I will leave it up to you get meaningful conclusions... all these are still pretty cryptic to me, I guess I need to read your book in order to get a better understanding ;-)

```
Bruno@bruno-laptop MINGW64 /c/Data/dev/02-forks/fast-json-stringify (feature/nullable-values)
$ yarn benchmark
yarn run v1.12.3
$ node bench.js
JSON.stringify array x 1,517 ops/sec ±13.85% (76 runs sampled)
fast-json-stringify array x 3,614 ops/sec ±6.05% (71 runs sampled)
fast-json-stringify-uglified array x 2,996 ops/sec ±6.91% (62 runs sampled)
JSON.stringify long string x 7,549 ops/sec ±5.85% (74 runs sampled)
fast-json-stringify long string x 7,987 ops/sec ±4.97% (76 runs sampled)
fast-json-stringify-uglified long string x 5,139 ops/sec ±17.74% (51 runs sampled)
JSON.stringify short string x 2,943,868 ops/sec ±10.55% (72 runs sampled)
fast-json-stringify short string x 13,271,440 ops/sec ±7.58% (65 runs sampled)
fast-json-stringify-uglified short string x 13,551,264 ops/sec ±5.64% (72 runs sampled)
JSON.stringify obj x 476,339 ops/sec ±9.61% (66 runs sampled)
fast-json-stringify obj x 2,078,797 ops/sec ±6.41% (48 runs sampled)
fast-json-stringify-uglified obj x 1,201,055 ops/sec ±17.85% (42 runs sampled)
Done in 73.66s.
```

```
Bruno@bruno-laptop MINGW64 /c/Data/dev/02-forks/fast-json-stringify (master)
$ yarn benchmark
yarn run v1.12.3
$ node bench.js
FJS creation x 4,179 ops/sec ±5.70% (78 runs sampled)
JSON.stringify array x 1,908 ops/sec ±5.85% (80 runs sampled)
fast-json-stringify array x 3,720 ops/sec ±8.65% (71 runs sampled)
fast-json-stringify-uglified array x 4,429 ops/sec ±3.12% (86 runs sampled)
JSON.stringify long string x 7,069 ops/sec ±10.44% (68 runs sampled)
fast-json-stringify long string x 7,578 ops/sec ±8.81% (72 runs sampled)
fast-json-stringify-uglified long string x 8,346 ops/sec ±3.68% (79 runs sampled)
JSON.stringify short string x 3,146,884 ops/sec ±6.10% (74 runs sampled)
fast-json-stringify short string x 16,888,382 ops/sec ±4.78% (82 runs sampled)
fast-json-stringify-uglified short string x 15,526,335 ops/sec ±5.67% (75 runs sampled)
JSON.stringify obj x 994,735 ops/sec ±5.61% (77 runs sampled)
fast-json-stringify obj x 3,940,538 ops/sec ±3.43% (80 runs sampled)
fast-json-stringify-uglified obj x 3,767,417 ops/sec ±4.78% (76 runs sampled)
Done in 72.46s.
```

**Comment by mcollina** [comment]
You likely have something else running on your machine, as the numbers of JSON.stringify have 2x difference :/. Can you run again?

**Comment by gendronb** [comment]
I can do that, but guess you would have better results running the tests on your machine... Mine is loaded (DBs, Docker, ...) and not really fast.

**Comment by gendronb** [comment]
Does it look any better to you?

```
Bruno@bruno-laptop MINGW64 /c/Data/dev/02-forks/fast-json-stringify (master)
$ yarn benchmark
yarn run v1.12.3
$ node bench.js
FJS creation x 3,747 ops/sec ±8.98% (75 runs sampled)
JSON.stringify array x 2,154 ops/sec ±2.40% (87 runs sampled)
fast-json-stringify array x 4,629 ops/sec ±3.30% (87 runs sampled)
fast-json-stringify-uglified array x 4,737 ops/sec ±2.24% (89 runs sampled)
JSON.stringify long string x 9,681 ops/sec ±2.25% (89 runs sampled)
fast-json-stringify long string x 9,832 ops/sec ±1.18% (92 runs sampled)
fast-json-stringify-uglified long string x 9,700 ops/sec ±1.54% (89 runs sampled)
JSON.stringify short string x 3,811,684 ops/sec ±2.74% (86 runs sampled)
fast-json-stringify short string x 19,057,372 ops/sec ±2.98% (88 runs sampled)
fast-json-stringify-uglified short string x 18,722,532 ops/sec ±2.48% (85 runs sampled)
JSON.stringify obj x 1,209,829 ops/sec ±2.11% (90 runs sampled)
fast-json-stringify obj x 4,114,364 ops/sec ±3.21% (85 runs sampled)
fast-json-stringify-uglified obj x 4,518,309 ops/sec ±2.02% (88 runs sampled)
Done in 72.71s.

Bruno@bruno-laptop MINGW64 /c/Data/dev/02-forks/fast-json-stringify (feature/nullable-values)
$ yarn benchmark
yarn run v1.12.3
$ node bench.js
FJS creation x 4,179 ops/sec ±5.58% (79 runs sampled)
JSON.stringify array x 2,173 ops/sec ±2.17% (87 runs sampled)
fast-json-stringify array x 4,700 ops/sec ±3.49% (87 runs sampled)
fast-json-stringify-uglified array x 4,618 ops/sec ±2.18% (89 runs sampled)
JSON.stringify long string x 9,477 ops/sec ±2.44% (88 runs sampled)
fast-json-stringify long string x 9,806 ops/sec ±1.37% (92 runs sampled)
fast-json-stringify-uglified long string x 9,634 ops/sec ±1.82% (90 runs sampled)
JSON.stringify short string x 3,741,979 ops/sec ±2.64% (85 runs sampled)
fast-json-stringify short string x 18,907,810 ops/sec ±1.60% (87 runs sampled)
fast-json-stringify-uglified short string x 18,876,740 ops/sec ±1.58% (87 runs sampled)
JSON.stringify obj x 1,192,430 ops/sec ±2.41% (90 runs sampled)
fast-json-stringify obj x 4,247,000 ops/sec ±2.48% (86 runs sampled)
fast-json-stringify-uglified obj x 4,498,992 ops/sec ±1.93% (88 runs sampled)
Done in 72.28s.
```

**Comment by mcollina** [comment]
This is ok, yes.

Did you find out what were you running that was causing issues?

File: index.js [hunk 1]
```diff
@@ -70,22 +70,40 @@ function build (schema, options) {
 
   var main
 
+  // wrapper function for processing nullable values
+  var processValue = (schema, code, name, nonNullHandler) => {
+    code += `
+      function ${name} (input) {
+        if(input === null) {
```
Comment (mcollina) [inline index.js]: can you add a space before the parentheses?

File: index.js [hunk 2]
```diff
@@ -70,22 +70,40 @@ function build (schema, options) {
...
     case 'number':
-      main = $asNumber.name
+      main = schema.nullable ? '$main' : $asNumber.name
+      code = schema.nullable ? processValue(schema, code, main, $asNumber.name) : code
       break
     case 'boolean':
-      main = $asBoolean.name
+      main = schema.nullable ? '$main' : $asBoolean.name
+      code = schema.nullable ? processValue(schema, code, main, $asBoolean.name) : code
       break
```
Comment (mcollina) [inline index.js]: Given that these are only for standard types, should we avoid generating a single function for each property but rather use some functions that we add? Something like: $asBooleanNullable, $asNumberNullable, $asStringNullable, etc..

**Pull request merged**